Evolve docs, round one

otherjoel opened the pull request:
Still in progress. In general, I've built out a new outline and shoved everything that was there before down into an “original docs” section for reference. 

I copied some content from the old to the new for convenience, but that’s not to say I might not rewrite it later on.

In case it's more convenient than pulling and rebuilding all the docs, I've uploaded a snapshot copy of the rendered docs at <https://joeldueck.com/what-about/congame/> for easy viewing. As I add commits to this branch, I’ll note whenever I update that snapshot.

MarcKaufmann commented on congame-doc/introduction.scrbl:
```diff
@@ -1,37 +1,530 @@
+conscript/local}.}
+
+@item{In Conscript, you can use “@"@" notation” to intermingle code and text. This is explained
+further in @secref["Scribble_Syntax"].}
```
The secref is not working on the snapshot you provided, it shows a red link.

otherjoel replied:
Got it. I retitled that section at the last minute, will fix.

MarcKaufmann commented on congame-doc/introduction.scrbl:
```diff
@@ -1,37 +1,530 @@
+upload our study to a Congame server. Without it, the server will not be able to access the study
+bound to the @racket[simple-study] identifier. 
+
+@margin-note{It’s a good practice to add @racket[(provide _studyname)] following every
```
I usually do `provide` at the top or before the definitions. And I would remove the `(provide (all-defined-out))`, I prefer being explicit about this.

MarcKaufmann commented on congame-doc/introduction.scrbl:
```diff
@@ -1,37 +1,530 @@
+@racket[(provide (all-defined-out))] expression, which will cover all the studies defined in the
+file with a single line of code.}
+
+Combining all these snippets, the code should look like the below example. Go ahead and save it as
```
Notice that you used different indentation for different `md`s, which is fine. It is important to have the same level of indentation within a given `md`, otherwise you get surprises. It is probably worth adding a note --- TODO comment --- to highlight this potential issue here (should they mess up) and add an example with a broken `md` and screenshot to the tutorial, where the first line is indented 4 spaces and a later paragraph 6 spaces, which will turn it into a quote and break a bunch of stuff. Showing this example will prime them so that when (not if) they hit this bug, they know where to go to find out what the problem is.

otherjoel replied:
I debated whether I should use `md` or `html` here. The indenting concern (which I’d forgotten about) makes me suspect `html` would ultimately be less confusing for the quick intro, and mesh better with the Overview (which has a short explainer on HTML). What do you think, should we switch to `html` here, or just add the caveats/examples around indentation in `md` — or both? 

In any case I will make a note to add warnings and examples of this caveat to the `md` and `md*` functions in the module reference.

MarcKaufmann replied:
We'll stick with `md` despite this gotcha, students just find it so much easier than `html` and they come across that type of syntax in more places including Git. `html` is for later, when one needs to add classes and ids to elements.

MarcKaufmann commented on congame-doc/introduction.scrbl:
```diff
@@ -1,37 +1,530 @@
+submitted, set the value in the @racket[_first-name] variable to the whatever the user has entered
+in @racket[(input-text)]. }
+
+@item{The @racket[submit-button] expression inserts the button the user can click to submit all the
```
This is another common bug: students either forget the `submit-button` or they use `button`. The tutorial should have both mistakes and highlight what happens. In the first case it is easy: you can't continue. In the second, when you click the button, the next page gets loaded, but no data gets submitted. We should probably guard against this by checking that there is a `submit-button` within each form.

otherjoel replied:
> We should probably guard against this by checking that there is a submit-button within each form.

I agree that this would ideally be a compile-time error. But I'll add some content around this in the next iteration. If/when you do make it an error, we can just update the example; it will still be useful to help students understand how to fix the error.

MarcKaufmann replied:
I discussed it with Bogdan, and it would be an anti-feature to add this as an error. So we'll just highlight it - tutorial probably rather than quick intro - and deal with it further if and when it comes up in the future.

Bogdan highlighted that one might want a submit button only to appear after some waiting time, or to have some other trigger for the form submission rather than a button click. So requiring a submit button is not always justified.

MarcKaufmann reviewed (changes requested):
I went through most things, skimmed the intro, I will let the students find potential issues. Looks great overall, and I really like the nice touch on the terminal and keyboard visuals. I left some minor comments here and there, nothing major, but we should fix them before I merge.

If you think my comments make something worse, let me know, or in the case of documentation --- what info goes where --- overrule me, when it's clear I merely have a casual opinion on something you have more experience than me. I usually try to make clear whether something really should be changed, or I merely feel it's better. Same if something might take substantial time.

MarcKaufmann commented on congame-doc/introduction.scrbl:
```diff
@@ -1,37 +1,530 @@
+@subsection{Preparing the study: from @racketmodname[conscript/local] to @racketmodname[conscript]}
+
+@bold{Important:} before uploading the study to a Congame server, we need to change it to use 
+@code{#lang conscript} instead of @code{#lang conscript/local}.
```
It is no longer necessary to remove the "/local", you should test that, but we allow both "conscript" and "conscript/local" and strip the "/local".

otherjoel replied:
In that case, is there ever a reason why I as a study author would want to use `conscript` rather than `conscript/local`?

MarcKaufmann replied:
Currently, yes, since "conscript/local" is the one that is needed to run the correct #lang offline, which has a `preview` function. When we upload it, we literally strip the string "/local" and thereby trigger the correct #lang conscript to be run on the server.

However, yesterday we started implementing a dockerized offline solution that should replace conscript/local, because keeping `conscript/local' in synch with `conscript` is takes around 25-30% of development and has caused half the bugs in the last half year. Moving to docker and providing a thin command line utility to automate common workflows such as updating a local study will mean that we only have to maintain one version.

In short, I expect us to take all `conscript/local` stuff out in the next month, so don't invest time into it.

MarcKaufmann commented on congame-doc/introduction.scrbl:
```diff
@@ -1,37 +1,530 @@
+
+The @racketmodname[conscript/local] environment is only useful for local testing; it’s quick to use,
+but it doesn’t permanently record any of the data collected, and cannot differentiate between
+participants. Using @racketmodname[conscript] @mark{unlocks access to the full server’s participant
```
What #lang conscript does is that it requires a congame server with DB to be run as a study - locally or online - which stores participant responses. Not sure what "unlocks access" would mean, you need to have an account on a server to do it, but then you can run such a study.

otherjoel replied:
I've commented out the whole section about changing `conscript/local` to `conscript`, which includes this section. Once I test that removing `/local` isn't needed on HEAD then I’ll delete it entirely.

MarcKaufmann commented on congame-doc/introduction.scrbl:
```diff
@@ -1,37 +1,530 @@
+participants will see this same study instance and the @onscreen{Enroll} link on their dashboard,
+allowing them to complete the study as well.}
+
+Once you’re done with the study, @mark{navigate back to the dashboard somehow}. Then click on the
```
Hm, I see, you have to manually change the url to get to the home page. This is somewhat annoying, but we don't want - on purpose - to show the homescreen, since for most participants we don't want that. Would it be helpful if we displayed the home button for the dashboard for admin and researcher users?

otherjoel replied:
> Would it be helpful if we displayed the home button for the dashboard for admin and researcher users?

I think that would be great. For everyone else, perhaps the server could auto-add a message at the bottom of any step that transitions to itself that says “You can now close this browser tab/window” (or similar).

MarcKaufmann replied:
> For everyone else, perhaps the server could auto-add a message at the bottom of any step that transitions to itself that says “You can now close this browser tab/window” (or similar).

Is this a problem you've had where you got confused by having a button that just goes to itself? In real studies, such buttons should be taken out, so that the final page is simply the final page with whatever custom message the study creator wants.

MarcKaufmann commented on congame-doc/introduction.scrbl:
```diff
@@ -1,37 +1,530 @@
+Once you’re done with the study, @mark{navigate back to the dashboard somehow}. Then click on the
+@onscreen{Admin} → @onscreen{Instance 1} link.
+
+@mark{All the way at the bottom}, you’ll see a section titled @bold{Participants} — and you’ll be
```
Why the mark, are you unsure this is correct, or you think it is an anti-feature? Our admin page is definitely an evolved beast and ugly.

otherjoel replied:
As I was writing the tutorial I asked myself why the Participants should be at the bottom rather than at the top. There may well be a good reason for this (I can think of a couple), but I can easily imagine a newcomer wondering why they need to wade past all the other stuff to get to the one thing they are most interested in from the beginning. Very minor note; I will probably take it out now that you’ve seen it so readers don’t get confused.

MarcKaufmann replied:
Can you add making a list of suggested improvements to the admin page as something to look at during the next round? It is one of those pain points we primarily don't update because we got used to it and don't really know how to improve. I agree that putting participants at the top is better since it's the default reason for coming to the admin.

MarcKaufmann reviewed (commented):
I forgot to look for marks that you made, I now commented on those too.

otherjoel commented on congame-doc/concepts.scrbl:
```diff
@@ -0,0 +1,116 @@
+study participants proceed through those steps.
+
+A @deftech{study instance} is a discrete time when a @tech{study} is run on a particular
+server. A study instance has a beginning and an end date/time. Congame keeps separate sets of
```
I see what you mean. I had a concept of the time-bounded nature of an instance as being one thing that distinguishes instances, but now I see what you mean. I think the fix here is simply to strike that sentence.

otherjoel commented on congame-doc/concepts.scrbl:
```diff
@@ -0,0 +1,116 @@
+
+@section{Building Studies}
+
+Two ways to author studies, depending on whether you're running your own Congame server.
```
I have been doing this with the `@mark` function I added, which highlights text a dull yellow. There are a lot more instances where I would mark up this kind of text, but I also don’t want the result to be distracting or confusing for your readers.

What if, instead of using `@mark` for this purpose, I add a `@tktk` function (named from an [editing convention](https://en.wikipedia.org/wiki/To_come_(publishing))) that made the text light grey, surrounded it in brackets, and put some kind of small note in the margin?

MarcKaufmann replied:
Yes, that's perfect.

otherjoel commented on README.md:
```diff
@@ -66,13 +107,43 @@ subprocesses defined in the `Procfile` whenever it is run.
+
+### Installing study packages
+
+* The `congame-web` package's `info.rkt` needs to `(define congame-studies)` like so:
```
I think I see what you mean — I misinterpreted the original README — but in digging into it for my own certainty, I ran into some questions, see #156.

MarcKaufmann replied:
Interesting, I'll check #156 but resolve this for now.

otherjoel commented:
I've now added commits, and added a 👍 response to all the review comments that I believe are resolved by the additional commits. (In case you prefer to check each one and mark them resolved yourself.)

MarcKaufmann commented on congame-doc/congame-install.scrbl:
```diff
@@ -0,0 +1,40 @@
+variables defined in this file into the environment of the
+subprocesses defined in the @filepath{Procfile} whenever it is run.
+
+The app expects to be run behind an SSL terminated connection (for
```
You should not change this value directly in `app.rkt`, you only need to add the flags `CONGAME_WEB_DEBUG=x` and `CONGAME_IDENTITY_DEBUG=x` in the `.env` file and it should do the trick. Directly changing this parameter will cause trouble when pushing to the server (it's fine if it is only used locally).

otherjoel replied:
Ah I didn't realize that (updating the parameter was how we did it during my initial setup call). Given that these values are present in the `.env.default` file that ships with the master branch, we should remove this guidance altogether, and add a separate section about what to change for running the server in prod.

MarcKaufmann commented:
First merge complete.
